File: searching-sorting/cyclic-sort.cpp
#include<bits/stdc++.h>
using namespace std; 

//* For the Questions with Pattern: 1 to N. This swapping used when numbers are in range.
//! Here, we have only O(n) time Complexity, as the number of Comparisions are 2*N - 1

void cyclic_sort(vector<int> &v){
    int present_index = 0;
    while(present_index < v.size()){
        int correct_index = v[present_index] - 1; //Correct Index of Number at i is (that number) - 1;
        if(v[present_index] != v[correct_index]){
            swap(v[present_index], v[correct_index]);
        }
        else{
            present_index++;
        }
    }
}

int main() 
{
    vector<int> v = {3,5,2,1,4};
    cyclic_sort(v);  
    for(int &value: v) cout<<value<<" "; 
}

File: searching-sorting/cyclic-sort.py
# 
# *For the Questions with Pattern: 1 to N. This swapping used when numbers are in range.
# *Here, we have only O(n) time Complexity, as the number of Comparisions are 2*N - 1

def cyclic_sort(arr):
    present_index = 0
    while present_index < len(arr):
        correct_index_for_value_at_present_index = arr[present_index] -1
        
        if arr[present_index] == arr[correct_index_for_value_at_present_index]:
            present_index += 1
        else:
            temp = arr[present_index]
            arr[present_index] = arr[correct_index_for_value_at_present_index]
            arr[correct_index_for_value_at_present_index] = temp

arr = [3,5,2,1,4,7,6,-1,-4,23]
cyclic_sort(arr)
print(arr)